### PR TITLE
[CSL-50] Check that more than 50% of stake participated in seed selection

### PR DESCRIPTION
**Pull request by neongreen**
In this PR “50% of stake” means “50% of richmen's stake”. However, @pva701 said

> я во всех своих телегах про 749 брал больше половины всего стейка
> и пацики меня не хэйтили

so maybe someone wants to look at his code as well :trollface:

### CODE REVIEW

**Comment by pva701** [comment]
@neongreen 
>In this PR “50% of stake” means “50% of richmen's stake”.

did you get approved from scientists?

**Comment by neongreen** [comment]
@pva701 well, @gromakovsky said

> Мне кажется более правильным сравнивать со стейком, принадлежащим богачам

but I'll go ask scientists just in case

**Comment by neongreen** [comment]
Asked Bernardo, he said it's okay:

> Since we assume that honest users controlling more than 50% of the total stake are online we can interpret the "stakeholder majority" either way.

**Comment by neongreen** [comment]
> TODO: test

Currently I can't get it to actually output the `SSC couldn't compute seed:` part even when I run 6 nodes locally and kill 3 of them.

File: src/Pos/Ssc/GodTossing/Seed.hs [hunk 1]
```diff
@@ -64,19 +75,22 @@ calculateSeed (getCommitmentsMap -> commitments) openings lShares = do
...
     -- i.e. ones which are in mustBeRecovered)
     let recovered :: HashMap StakeholderId (Maybe Secret)
+        -- NB. magical mystery list monad is used here
         recovered = HM.fromList $ do
             -- We are now trying to recover a secret for key 'k'
             id <- toList mustBeRecovered
             -- We collect all shares that 'k' has sent to other nodes
             let decryptedShares :: [Share]
-                decryptedShares = concatMap toList $ mapMaybe (HM.lookup id) (toList shares)
+                decryptedShares = concatMap toList $
```
Comment (ChShersh) [inline src/Pos/Ssc/GodTossing/Seed.hs]: `concatMap` is `>>=` for list monad. So this piece of code can be rewritten usit `<-`

File: lrc/Pos/Lrc/DB/Seed.hs [hunk 2]
```diff
@@ -0,0 +1,46 @@
+-- | We want store each epoch's seed in the database. This module does that.
```
Comment (pva701) [inline lrc/Pos/Lrc/DB/Seed.hs]: >we want *to* store

File: src/Pos/Lrc/Worker.hs [hunk 3]
```diff
@@ -140,25 +143,48 @@ lrcDo getPeers epoch consumers tip = tip <$ do
...
+            seed <- sscCalculateSeed epoch richmen >>= \case
+                Right s -> do
+                    logInfo $ sformat
+                        ("Calculated seed for epoch "%build%
+                         " successfully") epoch
+                    return s
+                Left err -> do
+                    logWarning $ sformat
+                        ("SSC couldn't compute seed: "%build) err
+                    logWarning "Going to reuse seed for previous epoch"
```
Comment (pva701) [inline src/Pos/Lrc/Worker.hs]: maybe better `logInfo`?

File: lrc/Pos/Lrc/DB/Seed.hs (hunk 2 above)
Comment (gromakovsky) [inline lrc/Pos/Lrc/DB/Seed.hs]: I agree. Why not fix?

File: src/Pos/Ssc/NistBeacon.hs [hunk 4]
```diff
@@ -68,6 +67,9 @@ instance SscGStateClass SscNistBeacon where
     sscLoadGlobalState = pure ()
     sscRollbackU _ = pure ()
     sscVerifyAndApplyBlocks _ _ = pass
-    sscCalculateSeedQ =
-        pure . Right . coerce . ByteArray.convert @_ @ByteString .
-            Hash.hashlazy @SHA256 . encode
+    sscCalculateSeedQ i _ = do
+        let h :: ByteString
+            h = ByteArray.convert $ Hash.hashlazy @SHA256 (encode i)
+        -- TODO: not sure that we don't have to cut it
```
Comment (gromakovsky) [inline src/Pos/Ssc/NistBeacon.hs]: Cut what? Seed? Make it shorter? It should be 32 bytes here and all seeds are 32 bytes (maybe except genesis one).

File: lrc/Pos/Lrc/DB/Seed.hs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import           Pos.DB.Class      (MonadDB)
+import           Pos.DB.Error      (DBError (DBMalformed))
+import           Pos.Lrc.DB.Common (getBi, putBi)
+import           Pos.Lrc.Genesis   (genesisSeed)
+import           Pos.Util.Util     (maybeThrow)
+
+getSeed :: MonadDB m => EpochIndex -> m SharedSeed
+getSeed epoch =
+    maybeThrow (DBMalformed "Seeds part of LRC DB is not initialized") =<<
+    getBi (seedKey epoch)
```
Comment (gromakovsky) [inline lrc/Pos/Lrc/DB/Seed.hs]: I disagree. If I use `getSeed 100500100500` it doesn't mean that DB is malformed. In initialization we put it only for 0-th epoch. Why not return `Maybe` as it is done everywhere else?

File: src/Pos/Lrc/Worker.hs [hunk 6]
```diff
@@ -142,23 +145,23 @@ lrcDo getPeers epoch consumers tip = tip <$ do
...
+                -- We have to compute richmen before calculating the seed
+                -- because seed calculation relies on knowing richmen. We
+                -- give 'calcAndStoreSeed' as a callback to
+                -- 'richmenComputationDo' because it's easier than reading
+                -- richmen from the database (I couldn't find a function that
+                -- does it on a locked LRC database)
+                let simpleConsumer = lrcConsumerFromComponentSimple @RCSsc bvdMpcThd
+                let calcSeedConsumer = simpleConsumer {
+                        lcComputedCallback = calcAndStoreSeed }
+                richmenComputationDo epoch (calcSeedConsumer:consumers)
```
Comment (gromakovsky) [inline src/Pos/Lrc/Worker.hs]: I don't understand this part. When we calculate seed for the 10-th epoch we use data from the 9-th epoch. So we should use richmen from the 9-th epoch, don't we? In this case all this stuff can be somewhat simplified I suppose.

File: lrc/Pos/Lrc/DB/Seed.hs (hunk 2 above)
Comment (neongreen) [inline lrc/Pos/Lrc/DB/Seed.hs]: Huh, I was pretty sure I fixed it.

File: lrc/Pos/Lrc/DB/Seed.hs (hunk 5 above)
Comment (neongreen) [inline lrc/Pos/Lrc/DB/Seed.hs]: I copied it from `getIssuersStakes`:

```haskell
getIssuersStakes :: MonadDB m => EpochIndex -> m IssuersStakes
getIssuersStakes epoch =
    maybeThrow (DBMalformed "Issuers part of LRC DB is not initialized") =<<
    getBi (issuersKey epoch)
```

But okay, I will change it.

File: src/Pos/Ssc/NistBeacon.hs (hunk 4 above)
Comment (neongreen) [inline src/Pos/Ssc/NistBeacon.hs]: I was lazy to lookup seed length.

File: lrc/Pos/Lrc/DB/Seed.hs (hunk 5 above)
Comment (gromakovsky) [inline lrc/Pos/Lrc/DB/Seed.hs]: Ok, I was not attentive enough while I was reviewing old @pva701's PR.

File: src/Pos/Lrc/Worker.hs (hunk 6 above)
Comment (neongreen) [inline src/Pos/Lrc/Worker.hs]: Does “richmen from 9th epoch” mean “nodes who were rich at the beginning of 9th epoch” or “nodes who were rich at the end (after the last slot) of the 9th epoch”?

Comment (gromakovsky) [inline src/Pos/Lrc/Worker.hs]: “richmen from 9th epoch” means “richmen computed by LRC when 9 was passed as argument”. It uses distribution from `crucialSlot 9`, i. e. `{8, epochSlots - slotSecurityParam - 1}`.

When we compute seed for the 10-th epoch we want to use richmen from `lrc 9`.

When we launch LRC for `9`-th epoch we use stake distrubtion

Comment (neongreen) [inline src/Pos/Lrc/Worker.hs]: Ah, okay.

**Comment by neongreen** [comment]
Okay, I fixed it (and tested).

**Pull request merged**